File: unity/ProfitableCore/Schema/UnassignShipRoleResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/unassignShipRoleResult.ts's UnassignShipRoleResult
// = UnassignShipRoleSucceeded | UnassignShipRoleRejected discriminated
// union.
public abstract class UnassignShipRoleResult
{
    public abstract bool Unassigned { get; }
}

public sealed class UnassignShipRoleSucceeded : UnassignShipRoleResult
{
    public override bool Unassigned => true;
    public CrewMember UpdatedCrewMember { get; init; } = new();
}

public sealed class UnassignShipRoleRejected : UnassignShipRoleResult
{
    public override bool Unassigned => false;
    public string Reason { get; init; } = string.Empty;
}

File: src/data/types/unassignShipRoleResult.ts
import type { CrewMember } from "./crewMember.ts";

// Ship Crew Roles amendment, task follow-up: the > Unassign action
// ship.md's own gap list flagged as missing. Same Result-union shape as
// AssignShipRoleResult/DismissResult -- "nothing to unassign" is a normal
// business rejection a UI must handle gracefully, not a caller error.
export interface UnassignShipRoleSucceeded {
  unassigned: true;
  updatedCrewMember: CrewMember;
}

export interface UnassignShipRoleRejected {
  unassigned: false;
  reason: string;
}

export type UnassignShipRoleResult = UnassignShipRoleSucceeded | UnassignShipRoleRejected;
